keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,833 @@
+* Provide request buffering at the Kubernetes Service level
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Allow the HPA to scale from and to zero using `minReplicas: 0` when explicitly enabled with a flag.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1: Scale a heavy queue consumer on-demand
+
+As the operator of a video processing pipeline, I would like to reduce costs. While video processing is CPU intensive, it is not a latency sensitive workload. Therefor I want
+my video processing workers to only be created if there is actually a video to be processed and terminated afterwards.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Currently disabling HPA is possible by manually setting the scaled resource to `replicas: 0`. This works as the HPA itself could never reach this state itself.
+As `replicas: 0` is now a possible state when using `minReplicas: 0` it can no longer be used to differentiate between manually disabled or automatically scaled to zero.
+
+Additionally the `replicas: 0` state is problematic as updating a HPA object `minReplicas` from `0` to `1` has different behavior. If `replicas` was `0` during the update, HPA
+will be disabled for the resource, if it was `> 0`, HPA will continue with the new `minReplicas` value.
+
+To resolve this issue the KEP is introducing an explicit `enableScaleToZero` property to explicitly enable/disable scale from/to zero.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+From an UX perspective the two stage opt-out / opt-in from scale to zero might feel a bit tedious, but the only other available option seems to be deprecating the implicit HPA pause on `replicas: 0`. While this might provide an improved
+UX, it would require a full deprecation cycle (12 months) before graduating this feature from alpha to beta.
```

Comment:
Removing the HPA pause on `replicas: 0` only for HPAs with `minReplicas: 0` makes sense to me. That's an opt-in on the HPA itself, would be documented as part of `scale-to-zero`, no existing HPAs (other than ones using the alpha scale-to-zero feature) would change behavior, and no 12-month deprecation is required.